### ISSUE TITLE
Running on large codebase is not parallelised

### ISSUE THREAD

**Issue by muglug** [VKCOM/noverify#333]
This is the second run with `-cache-dir` for the given coebase

```
time $GOPATH/bin/noverify -exclude='vendor/|lib/' -cache-dir=/tmp/noverify .
2020/01/11 00:31:31.928425 built on: 20200109230528 OS: Matthews-MacBook-Pro-2.local x86_64 Commit: 50e94a2865362c7fd73de7e24c4082e56abe2400
2020/01/11 00:31:31.928605 Started
2020/01/11 00:31:32.286324 Indexing [.]
2020/01/11 00:32:03.865122 Failed parsing <PATH>/vendor/composer/composer/tests/Composer/Test/Autoload/Fixtures/classmap/StripNoise.php: Empty root node
2020/01/11 00:32:17.675855 Linting
MAYBE   phpdoc: Missing PHPDoc for "count" public method at <PATH>/base/Collection.php:59
    public function count()
                    ^^^^^
...

2020/01/11 00:33:04.622640 Found 159350 critical reports

real    1m32.775s
user    1m11.121s
sys 0m53.554s
```

The first error is wrong, btw (as are the vast majority of the 159,350). According to PHPDoc standards, the docblock for `count` is assumed inherited from `ArrayObject::count`.

**Comment by YuriyNasretdinov** [comment]
Regarding the parallelism:

1. Do you have GOMAXPROCS environment variable set by any chance? If so, please unset it or set to the number of logical cores in your system
2. How fast is the storage for the source code and for the cache directory? Is it SSD, HDD, a network drive?
3. What operating system you are running?
4. Which Go version?

I am sure that it's the latest Go version, SSD and no GOMAXPROCS overrides, but if not, it can make a lot of difference.

Thanks

**Comment by muglug** [comment]
- I'm using an SSD (1TB flash storage)
- no `GOMAXPROCS` override set
- running OS X 10.14.6
- go version go1.13.6 darwin/amd64

Again, I see parallelism when I run on Psalm's source-code (which is open-source):

```
real	0m2.036s
user	0m9.908s
sys	0m0.692s
```

But not when running on Vimeo's.

**Comment by YuriyNasretdinov** [comment]
So, given the fact that noverify generates 160k warnings I assume that it actually takes most of the time. You can check that it is indeed the case by only allowing certain checks by using -allow-checks=... . It will still analyse everything but it will only print the selected types of warnings.

**Comment by YuriyNasretdinov** [comment]
So even on psalm codebase only allowing certain (e.g. non-existent checks) to be printed improves performance significantly:

```
$ ./psalm # no pcntl on my machine sadly so let's just compare user+sys time
...
       29.33 real        28.49 user         0.81 sys
```

Several different noverify runs (not sure about settings used in Psalm so just in case):

```
$ time noverify -cache-dir=$HOME/tmp/noverify -cores=4 .
...
2020/01/15 23:52:36.367306 Found 29149 critical reports
        3.49 real        12.95 user         1.40 sys
$ time noverify -cache-dir=$HOME/tmp/noverify -exclude=vendor/ -cores=4 . 
...
2020/01/15 23:53:09.711542 Found 16839 critical reports
        2.20 real         7.90 user         0.99 sys
$ time noverify -cache-dir=$HOME/tmp/noverify -exclude=vendor/ -allow-checks=fatal -cores=4 .
...
2020/01/15 23:53:30.505898 Linting
        1.74 real         7.81 user         0.82 sys
```

So in the worst-case noverify spends about ~half of CPU time of Psalm which is not that impressive at all for a linter written in Go instead of PHP.
If we exclude `vendor/` however it starts to consume much less CPU (~8 seconds instead of ~29 seconds).

Note that even in the runs in the local console there is some measurable overhead for printing all 16k results (extra 500ms of real time for printing 16k, so it would be 5s for 160k). I assume that you run the vimeo codebase analysis on a remote server so you're probably limited by network and/or your terminal in this case.

**Comment by muglug** [comment]
> I assume that you run the vimeo codebase analysis on a remote server so you're probably limited by network and/or your terminal in this case.

No, the Vimeo codebase is relatively small (~700K LOC excluding `vendor`) so the codebase can happily fit on disk, which is where it's analysed by Psalm.

Running Psalm on Vimeo's codebase with multiple processes gives this output:

```php
real	1m27.694s
user	6m31.718s
sys	0m36.477s
```

Here Psalm is ignoring around 26K issues, so behaviour shouldn't be too dissimilar.

**Comment by YuriyNasretdinov** [comment]
Did running with e.g.  -allow-checks=fatal change anything in terms of performance though?

**Comment by YuriyNasretdinov** [comment]
Also from the noverify logs it looks like the indexing takes a long time too:

```
2020/01/11 00:31:32.286324 Indexing [.]
2020/01/11 00:32:03.865122 Failed parsing <filename>
2020/01/11 00:32:17.675855 Linting
```

Indexing took ~45 seconds which may indicate that the cache was empty?

The actual linting step took roughly the same amount of time:

```
2020/01/11 00:32:17.675855 Linting
...
2020/01/11 00:33:04.622640 Found 159350 critical reports
```

So linting actually took ~47 seconds to complete.

Also system time looks very high to me:

```
user    1m11.121s
sys 0m53.554s
```

Usually it's not that high, e.g. for psalm codebase:

```
7.90 user         0.99 sys
```

Here the system time and the user time are not 1:8 but closer to 3:4.
The high system% can be caused by many different things, but that is a strange symptom to me. From the top of my head it's hard to tell what might be causing it, but there are some of my guesses:

1. You have some very big files with php extensions that are actually not PHP code: php,inc,php5,phtml,inc
2. You have a lot of files in excluded directory. It should only affect the "indexing" phase as it indexes all files regardless of -exclude filter
3. Not enough RAM to parse some files? Go parser consumes a significant amount of memory when doing analysis in parallel. There are some limits that you can set up to (indirectly) limit memory usage by setting -max-sum-filesize=XXX bytes. The default is 20 MiB total file size that should be enough for most use cases but might cause memory usage go up to 5-6 GiB during parsing.
4. Something else? There are -pprof=... and -cpuprofile=... and -memprofile=... flags that can be used to profile the bottlenecks. Most certainly making several full goroutine dumps from the running process by accessing -pprof=... webserver would be very useful to spot parallelism bottlenecks.

Thank you!

**Comment by muglug** [comment]
With `-allow-checks=fatal`:

```
2020/01/16 09:35:56.640120 built on: 20200109230528 OS: Matthews-MacBook-Pro-2.local x86_64 Commit: 50e94a2865362c7fd73de7e24c4082e56abe2400
2020/01/16 09:35:56.640374 Started
2020/01/16 09:35:56.970212 Indexing [.]
2020/01/16 09:36:29.057180 Failed parsing vendor/composer/composer/tests/Composer/Test/Autoload/Fixtures/classmap/StripNoise.php: Empty root node
2020/01/16 09:36:44.939026 Linting

real	1m31.777s
user	1m16.313s
sys	0m52.575s
```

**Comment by YuriyNasretdinov** [comment]
Yeah, when I noticed that the indexing phase took 45 seconds it probably indicates that the reports printing is not actually the underlying issue, you're right.

**Comment by muglug** [comment]
But yes, it's strange that indexing (once cached) takes so much time.

Psalm's scanning behaviour is a little different:
1. it starts with a [list of directories](https://github.com/vimeo/psalm/blob/26bf1834304cbd5ca86187009a1cf39f471ce26c/psalm.xml.dist#L19-L21) it scans, and collects all referenced class, interface and trait names
2. then uses the Composer classloader to determine where any as-yet-unscanned files are (all of those should be in `vendor`), collecting any new referenced class, interface and trait names
3. repeat 2. until all referenced classes, traits and interfaces were scanned
4. calculate inheritance
5. analyse files in config directories

Steps 1-3 are parallelised and cached, 4 always occurs.

Importantly, this means that only a subset of files in `vendor` are ever scanned, so Psalm doesn't waste time.

**Comment by YuriyNasretdinov** [comment]
Yeah, so in noverify the process of walking though the directories is not parallelised but the actual parsing and indexing of the files that that were found is parallel. Noverify couldn't really use PHP autoload (or, rather, it probably could but we don't want to). So unless there are tons of non-PHP files in vendor/ and lib/ the process should be parallel.

Could you please count the number of files in excluded directories just to be able to understand whether or not that can be the root cause? Also running -pprof=... and making several full goroutine dumps in the process would also help to find the root cause.

Thanks.

**Comment by muglug** [comment]
btw

> You have some very big files with php extensions that are actually not PHP code: php,inc,php5,phtml,inc
nope, they're all PHP (though the phtml files have a lot of HTML too)

> You have a lot of files in excluded directory. It should only affect the "indexing" phase as it indexes all files regardless of -exclude filter

from `cloc vendor`:

```
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
PHP                          20140         255784         681107        1471616
```

> Not enough RAM to parse some files? 

Laptop has 16G of RAM

**Comment by YuriyNasretdinov** [comment]
> Laptop has 16G of RAM
Yes, if you don't have big files (like ~20 MiB in size) noverify should consume ~1-2 GiB of RAM during parsing using 4-8 cores. You can check how much memory is actually consumed in Activity Monitor.

**Comment by YuriyNasretdinov** [comment]
> from cloc vendor:

1-2m SLOC should not be a problem when indexing (the indexing speed on 4-core i7 should be around ~1m sloc/sec).

What can really affect the indexing and the parsing speed is the number of other files. Maybe a ton of .git files or something like this? Thanks

**Comment by muglug** [comment]
I don't have time to really dig into this right now but, at the very least, I think noverify doesn't need to scan all those `vendor` files.

I just ran Psalm on the same codebase, and it only scans 2,257 PHP files in `vendor`. It would be reasonably trivial for noverify to first parse Composer's generated autoload files to understand the mapping between class name & file name (though if VK doesn't use Composer it would be much more difficult).

**Comment by YuriyNasretdinov** [comment]
What I'm saying is that during the indexing phase it's completely fine in terms of performance to read even ~2m SLOC provided that it's just PHP files that are there. If noverify was just indexing all PHP files from vendor in a single thread it would still only take ~8 seconds for ~2m SLOC, not 45 sec. And indexing is parallelised too so it should really take 2s to read all 20000 PHP files from vendor. So it shouldn't be the problem we're looking for.

Ok, I will assume that we need to speed up directory traversal (maybe?) but I'd like to have some more proofs of that it is an actual issue :).

**Comment by YuriyNasretdinov** [comment]
I have implemented a small feature (https://github.com/VKCOM/noverify/pull/334) so that noverify can print which files took longer than X to parse.

With that pull request you should be able to run the following:

```
$ time $GOPATH/bin/noverify -exclude='vendor/|lib/' -cache-dir=/tmp/noverify -debug-parse-duration='500ms' .
```

This should show files that took more than 500ms to parse. On Psalm codebase it showed the following:

```
Parsing of src/Psalm/Internal/Visitor/ReflectorVisitor.php took 115.846202ms
Parsing of src/Psalm/Internal/CallMap.php took 323.044863ms
```

ReflectorVisitor.php is not that big (140 Kb) but it has a lot of nested if statements that cause a lot of context cloning.
CallMap.php is ~1 Mb in size that it seems that it is actually excluded from the Psalm config. It does not contribute too much to the overall execution time for noverify though.

**Comment by YuriyNasretdinov** [comment]
I highly doubt that there are such huge files in Psalm repository but it's worth checking and you don't need to do any pprof dumps for that.

**Comment by YuriyNasretdinov** [comment]
I have also written a smallish script to automate taking of pprof dumps:

```
#!/bin/sh
time $GOPATH/bin/noverify -exclude='vendor/|lib/' -cache-dir=/tmp/noverify -debug-parse-duration='500ms' -pprof=localhost:8080 -allow-checks=fatal . &
mkdir -p pprof

for i in {1..1000}; do
	sleep 5

	# printing time with milliseconds in macOS without external tools is hard :)
	php -n -r 'echo (new DateTime)->format("Y/m/d H:i:s.u") . " Getting profile '$i'" . PHP_EOL;'

	if ! curl -s 'http://localhost:8080/debug/pprof/goroutine?debug=2' >pprof/$i.txt; then
		rm pprof/$i.txt
		wait
		exit
	fi
done
```

It writes files to pprof/$i.txt every 5 seconds. This is an example run on some ~4m SLOC codebase (not VK):

```
$ ./pprof.sh                                                                                                                                                                                                                  2 ms  
2020/01/16 22:15:43.782736 built on: 20200116213012 OS: Mac-Pro-Mac-2.local x86_64 Commit: 50e94a2865362c7fd73de7e24c4082e56abe2400
2020/01/16 22:15:43.782951 Started
2020/01/16 22:15:44.143679 Indexing [.]
2020/01/16 22:15:48.873492 Getting profile 1
2020/01/16 22:15:50.509058 Linting
2020/01/16 22:15:54.055449 Getting profile 2
2020/01/16 22:15:59.197229 Getting profile 3
2020/01/16 22:16:04.332843 Getting profile 4

real	0m22.546s
user	2m35.413s
sys	0m9.873s
2020/01/16 22:16:09.423017 Getting profile 5
```

Here is a sample output.

In 1.txt you can see:
- `linter.ReadFilenames` goroutine that is busy reading some directory
- a couple `linter.readMetaCache` goroutines that are reading cache
- several `linter.ParseFilenames` in [chan receive] state that indicates that they are waiting for new files to parse (waiting for ReadFilenames goroutine actually)

In 2.txt and 3.txt you can see:
- `linter.ReadFilenames` goroutine that is busy reading some directory 
- very few `linter.ParseFilenames` goroutines in [chan receive] state waiting for ReadFilenames
- a lot of `linter.ParseContents` goroutines doing file parsing or analysis

In 4.txt which is almost an end:
- `linter.ReadFilenames` goroutine that is busy reading some directory
- all 8 `linter.ParseFilenames` in [chan receive] state waiting for more files to parse

So, in conclusion, even in "ideal" conditions the bottleneck can actually be directory traversal, weirdly enough. I think it is worth fixing anyway.

[1.txt](https://github.com/VKCOM/noverify/files/4074017/1.txt)
[2.txt](https://github.com/VKCOM/noverify/files/4074018/2.txt)
[3.txt](https://github.com/VKCOM/noverify/files/4074019/3.txt)
[4.txt](https://github.com/VKCOM/noverify/files/4074020/4.txt)

**Comment by YuriyNasretdinov** [comment]
So with https://github.com/VKCOM/noverify/pull/335 (I replaced filepath.Walk with faster https://github.com/karrick/godirwalk Walk) you should theoretically see up to 10x improvement in directory traversal speed on Darwin. It didn't affect times for the repository that I used but the pull request ensures that the directory traversal is not blocked by parsing any more and it uses much faster traversal code so if this doesn't help I'm very curious to see the pprof.

**Comment by muglug** [comment]
That seemed to do the trick:

```
real	0m28.935s
user	1m8.864s
sys	0m21.196s
```

[1.txt](https://github.com/VKCOM/noverify/files/4074409/1.txt)
[2.txt](https://github.com/VKCOM/noverify/files/4074410/2.txt)
[3.txt](https://github.com/VKCOM/noverify/files/4074411/3.txt)
[4.txt](https://github.com/VKCOM/noverify/files/4074412/4.txt)
[5.txt](https://github.com/VKCOM/noverify/files/4074413/5.txt)

**Comment by YuriyNasretdinov** [comment]
I wouldn't close this issue just yet as in the case above noverify managed to utilise ~7.5 logical cores out of 8. According to the results that you presented average parallelism is (68+21)/29 = ~3 cores out of 8, which is still very far from perfect.

Also from the profiles you shared as far as I can tell noverify is still mostly waiting on filenames from a directory reading goroutine. Which means that for codebases like Vimeo that have a lot of non-PHP files that need to be skipped we need to be able to run directory traversal in parallel too.

**Comment by YuriyNasretdinov** [comment]
muglug@, could you please also test https://github.com/VKCOM/noverify/pull/336 and see if the results become even better? Thanks

**Comment by muglug** [comment]
Unfortunately I won’t have access to a computer until Monday evening, but I will then.

**Comment by YuriyNasretdinov** [comment]
Thank you!

**Comment by muglug** [comment]
```
2020/01/20 22:27:09.117895 Found 154706 critical reports

real	0m18.017s
user	1m13.810s
sys	0m40.321s
```

Looks like that'll fix it, albeit with a few errors related to non-PHP files, where a symlink could not be resolved:

```
Error while reading directory "packages/internal_package_name/node_modules/.bin": stat packages/internal_package_name/node_modules/.bin/prettier: no such file or directory
```

Here `packages/internal_package_name/node_modules/.bin/prettier` is a symbolic link to a file that doesn't actually exist.

It didn't affect the analysis though (the same 160,000 issues were emitted).

**Comment by YuriyNasretdinov** [comment]
> Looks like that'll fix it, albeit with a few errors related to non-PHP files, where a symlink could not be resolved:
Yeah, I guess we also need to add functionality to be able to ignore dirs that definitely don't contain any PHP-code like `node_modules` and that don't need to be traversed at all :).
So I don't think we'll merge the pull request with directory traversal parallelisation then because even though it speeds things up a little in your case, in general is has the opposite effect because parallel directory traversal introduces measurable overhead (especially in sys time, it was 20s before and now it's 40s).

**Comment by YuriyNasretdinov** [comment]
https://i.redd.it/tfugj4n3l6ez.png

**Comment by YuriyNasretdinov** [comment]
![](https://i.redd.it/tfugj4n3l6ez.png)

**Comment by muglug** [comment]
>  even though it speeds things up a little in your case, in general it has the opposite effect

On which codebases? I imagine there are many monorepos where PHP files are a subset of the total.

**Comment by YuriyNasretdinov** [comment]
> On which codebases? I imagine there are many monorepos where PHP files are a subset of the total.

I guess you're right in that it really depends on the project. Most open-source PHP codebases, for example, don't have huge directories like `node_modules` that don't contain PHP files at all that lead to stalls in the parsing. I have also worked on a couple relatively big PHP codebases and  even though `node_modules` was present it didn't affect `noverify` performance all that much (I tested on Linux though which is like ~5x faster in terms of directory traversal speed in Go).

**Comment by YuriyNasretdinov** [comment]
Actually, thinking about the original problem... You can tell noverify to only analyse certain directories in the same fashion that you do with Psalm:

instead of running `noverify -exclude='vendor/|lib/' -cache-dir=/tmp/noverify .` you can run something like this:

`noverify -exclude='vendor/|lib/' -cache-dir=/tmp/noverify src/ lib/ vendor/`

If you only have PHP code in those directories, it should be closer to apples-to-apples comparison with Psalm in this case.

**Comment by muglug** [comment]
> instead of running `noverify -exclude='vendor/|lib/' -cache-dir=/tmp/noverify .` you can run something like this

Oh ok! Where does that leave this ticket?

**Comment by YuriyNasretdinov** [comment]
> Oh ok! Where does that leave this ticket?

I will think (and probably discuss) how to make sure we don't traverse directories we don't need in a more straightforward way (e.g. ignore `node_modules` by default but allow to change that). Also it is probably worth documenting that it is possible to specify multiple directories (or individual files for that matter) to analyse.

I would still like the default mode of operation be efficient enough that you shouldn't have to worry about this in 99% of cases.

So I would like to leave the issue open until we address the issue when there is a huge `node_modules` (or any other) directory that takes a lot of time to traverse and leads to processing stalls.

**Comment by muglug** [comment]
btw running this at work (slightly faster machine) with those directory filters:

```
time ~/go/bin/noverify -exclude='vendor/|lib/' <directories and files>
2020/01/21 15:35:48.766695 built without version info (try using 'make install'?)
2020/01/21 15:35:48.766883 Started
2020/01/21 15:35:48.893931 Indexing [base routes definitions objects controllers_v6 src scripts tests cron views_v6 web/vimeo.index.php]
2020/01/21 15:35:50.384088 Linting

< many many errors >

real	0m6.723s
user	0m44.449s
sys	0m2.245s
```

**Comment by YuriyNasretdinov** [comment]
So the expected numbers for running on mid-range 4-core Core i7 are around ~1M sloc/sec of indexing  speed and ~100K sloc/sec of analysis speed. In your case (800K sloc, right?) The numbers seem to be the following:

1) indexing took 1.5 seconds, so I assume it would mean around 500k sloc/sec for indexing
2) everything else took 4.2 seconds — around 200k sloc/sec analysis speed.

So the numbers are a bit off but in general this is much closer to the level of performance that I expected than previously.

Thank you for your analysis!

**Comment by YuriyNasretdinov** [comment]
UPD: I think I miscalculated a bit:

1. Indexing did take around 1.5 seconds for 700k sloc which gives around ~450k sloc/sec indexing speed which is lower than I expected. Probably also worth investigating in the future (indexing is more sensitive to directory traversal time so maybe we still want to have paralell traversal after all)
2. Everything else took 5.2 seconds, not 4.2, and the codebase is a bit smaller so it gives ~166k sloc/sec which is a bit higher than expected.

Also you still need to specify `vendor/` and `lib/` in the arguments or otherwise those directories won't be indexed (they won't be analysed as you excluded them but the indexing part is needed so that you don't get errors for the usage of classes/functions/etc that are defined in these directories).
